[Turnstile] Tutorial for conditionally enforcing Turnstile with HTMLRewriter (#21311)
cloudflare/cloudflare-docs: src/content/docs/turnstile/tutorials/conditionally-enforcing-turnstile.mdx

---
title: Conditionally enforce Turnstile
pcx_content_type: tutorial
updated: 2025-04-01
difficulty: Intermediate
languages:
  - TypeScript
tags:
  - Node.js
sidebar:
  order: 6
---

This tutorial explains how to conditionally enforce Turnstile based on the incoming request, such as a pre-shared secret in a header or a specific IP address.

## Overview

> Occasionally, you might have setups such as automation that cannot load or run the Turnstile challenge. This tutorial will demonstrate using [`HTMLRewriter`](/workers/runtime-apis/html-rewriter/) to conditionally handle the [client-side widget](/turnstile/get-started/client-side-rendering/) and [siteverify API](/turnstile/get-started/server-side-validation/) when specific criteria are met.

Reviewer: ```suggestion
You may have setups such as automation that cannot load or run the Turnstile challenge. Using [`HTMLRewriter`](/workers/runtime-apis/html-rewriter/), this tutorial will demonstrate how to conditionally handle the [client-side widget](/turnstile/get-started/client-side-rendering/) and [siteverify API](/turnstile/get-started/server-side-validation/) when specific criteria are met.
```

Whilst this tutorial demonstrates removing the Turnstile client-side elements when criteria are met, you could instead conditionally insert them if you prefer.

:::caution

It is critical to make sure you are validating tokens with the siteverify API when your criteria for enforcing Turnstile are not met.

> It is not sufficient to just remove the client-side widget from the page, as an attacker could simply forge the request to your API.

Reviewer: ```suggestion
It is not sufficient to only remove the client-side widget from the page, as an attacker can forge the request to your API.
```

---
title: Conditionally enforce Turnstile
pcx_content_type: tutorial
updated: 2025-04-01
difficulty: Intermediate
languages:
  - TypeScript
tags:
  - Node.js
sidebar:
  order: 6
---

This tutorial explains how to conditionally enforce Turnstile based on the incoming request, such as a pre-shared secret in a header or a specific IP address.

## Overview

You may have setups such as automation that cannot load or run the Turnstile challenge. Using [`HTMLRewriter`](/workers/runtime-apis/html-rewriter/), this tutorial will demonstrate how to conditionally handle the [client-side widget](/turnstile/get-started/client-side-rendering/) and [siteverify API](/turnstile/get-started/server-side-validation/) when specific criteria are met.

> :::note
> While this tutorial removes Turnstile client-side elements when specific criteria are met, you could instead conditionally insert them.
> :::
> :::caution
> 
> It is critical to make sure you are validating tokens with the siteverify API when your criteria for enforcing Turnstile are not met.
> 
> It is not sufficient to only remove the client-side widget from the page, as an attacker can forge the request to your API.
> 
> :::

Reviewer: ```suggestion
:::note
While this tutorial removes Turnstile client-side elements when specific criteria are met, you could instead conditionally insert them.
:::

:::caution
It is critical to make sure you are validating tokens with the siteverify API when your criteria for enforcing Turnstile are not met.

It is not sufficient to only remove the client-side widget from the page, as an attacker can forge the request to your API.
:::
```
